Add an assert_not_impl macro

197g opened the pull request:
This is the polar opposite of the assert_impl macro and abuses name
resolution with trait.

A trait with type parameter `A` provides a simple method. The trait is
implemented on the checked type `T` potentially twice: with type
parameter unit `A=()` for all types `T`, and with another parameter type
-abitrarily a new `Invalid`- when the `T` implements the trait bounds.
We then try to select `T` as an instance of the trait, using turbo-fish
to try and access one of its items, without providing the trait type
parameter. When only the first instance is implemented the type checker
can infer `A=()` but when the second is implemented as well it can not.

No other ambiguity can occur, only the trait's type parameter is
ambiguous. Thus, type inference and compilation succeeds if the trait
bound is not satisified by the input type.

197g committed:
Add an assert_not_impl macro

This is the polar opposite of the assert_impl macro and abuses name
resolution with trait.

A trait with type parameter `A` provides a simple method. The trait is
implemented on the checked type `T` potentially twice: with type
parameter unit `A=()` for all types `T`, and with another parameter type
-abitrarily a new `Invalid`- when the `T` implements the trait bounds.
We then try to select `T` as an instance of the trait, using turbo-fish
to try and access one of its items, without providing the trait type
parameter. When only the first instance is implemented the type checker
can infer `A=()` but when the second is implemented as well it can not.

No other ambiguity can occur, only the trait's type parameter is
ambiguous. Thus, type inference and compilation succeeds if the trait
bound is not satisified by the input type.

iliekturtles commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,115 @@
+#[macro_export(local_inner_macros)]
+macro_rules! _assert_not_impl {
+    ($x:ty, $($t:path),+ $(,)*) => {
+        {
```
Is this block necessary since this macro arm is already nested inside `fn $label() { ... }` from the other macro arm?

HeroicKatora replied:
I had simply taken the style from `assert_impl!` but it probably is not strictly necessary.

nvzqz replied:
I'm not familiar enough with the macro hygiene rules to know how this would be affected by not having a scope so I'd like to keep this scoped just for sanity.

iliekturtles reviewed (commented):
This is exciting!

HeroicKatora commented:
I needed to modify the design slightly, as the method call was possibly not completely unambiguous. Now the ambiguity has moved to the type parameter of a single locally defined trait which allows us to restrict the set of possibly applicable methods to that trait alone using turbo-fish syntax. By leaving out the trait's type parameter it can only be inferred when one impl is applicable.

HeroicKatora commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,107 @@
+macro_rules! _assert_not_impl {
+    ($x:ty, $($t:path),+ $(,)*) => {
+        const _: fn() -> () = || {
+            struct Check<T: ?Sized>(T);
```
It may be possible to replace this with `_core::marker::PhantomData<T>`. Would that be preferable?

nvzqz replied:
I think that `Check` gets the point across sufficiently.

nvzqz commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,107 @@
+/// [`Sync`]: https://doc.rust-lang.org/std/marker/trait.Sync.html
+/// [blanket]: https://doc.rust-lang.org/book/second-edition/ch10-02-traits.html#using-trait-bounds-to-conditionally-implement-methods
+#[macro_export(local_inner_macros)]
+macro_rules! assert_not_impl {
```
Please rename this to `assert_not_impl_all!` and add a `assert_not_impl_any!` counterpart.

For `assert_not_impl_any!`, I strongly suggest that the implementation does:

```rust
$(impl<T: ?Sized + $t> AmbiguousIfImpl<&dyn $t> for Check<T> {})+
```

nvzqz commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,107 @@
+            trait AmbiguousIfImpl<A> { fn some_item() { } }
+
+            impl<T: ?Sized> AmbiguousIfImpl<()> for Check<T> { }
+            impl<T: ?Sized $(+ $t)*> AmbiguousIfImpl<u8> for Check<T> { }
```
Rather than use `u8` here, add `struct Invalid;` under `Check` (or a better name if you think of one). Since it never gets constructed, make sure to add `#[allow(dead_code)]` to the declaration.

For `assert_not_impl_any!`, you should implement this bit like the following. Notice that each `Invalid` type is unique because of the scoping.
```rust
$({
    #[allow(dead_code)]
    struct Invalid;

    impl<A: ?Sized + $t> AmbiguousIfImpl<Invalid> for Check<A> {}
})+
```

Also, you should be matching multiple `$t` via `+` and not `*`.

nvzqz commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,107 @@
+            impl<T: ?Sized> AmbiguousIfImpl<()> for Check<T> { }
+            impl<T: ?Sized $(+ $t)*> AmbiguousIfImpl<u8> for Check<T> { }
+
+            <Check::<$x> as AmbiguousIfImpl<_>>::some_item()
```
Please do `let _ = <Check::<$x> as AmbiguousIfImpl<_>>::some_item;`.

nvzqz commented on src/assert_not_impl.rs:
````diff
@@ -0,0 +1,107 @@
+/// ```
+///
+/// Note that all traits given in a single invocation need to not be implemented. If you want to
+/// check that none of multiple trait are implemented you must invoke the macro several times.
````
Change this to refer to `assert_not_impl_any!` instead with a working local docs link. Also move this comment to right before "Examples".

nvzqz commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,107 @@
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // Cell<u32> is not both Sync and Send
```
Make this normal docs text.

nvzqz commented on src/assert_not_impl.rs:
```diff
@@ -0,0 +1,107 @@
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // But it is Send, this fails to compile
```
Make this normal docs text. For visual purposes, I don't like having code blocks separated without anything in between to pad them.

nvzqz reviewed (changes requested):
Thank you so much for this!! I was previously convinced that this wasn't possible. I love cursed hacks like these. ❤

Regarding style, please remove whitespace between empty braces.

I'll be more than happy to merge this PR once you've made my requested changes. Also, the changes in this PR should be detailed in `CHANGELOG.md` under "Unreleased" in a new "Added" section.

nox commented:
Out of curiosity, why do `Check<T>` and the type parameter on the trait exist at all?

To check that a type `S` doesn't implement a trait `T`, one can do:

```rust
trait ShouldNotImplT {}
impl ShouldNotImplT for U where U: T {}
impl ShouldNotImplT for S {}
```

HeroicKatora commented:
@nox No particular reason, it turns out, only that a previous version of the implementation had a `self` parameter instead of the turbofish for trait resolution. Thanks for the simplification suggestion.

197g committed:
Add assert_not_impl_any

Makes use of the same trick as assert_not_impl_all but adding multiple
possible implementations of the trait instead of at most two.

nvzqz commented:
Awesome, thank you very much!

nvzqz commented:
This closes https://github.com/nvzqz/static-assertions-rs/issues/8.

WorldSEnder commented:
Coming back to this a few years later, the `self` parameter can be handy if you have a value of a type you can't name, e.g. as happening with futures.

```rust
async fn some_lib_feature() {
    // Do some async work. To future proof this, and not cause unintentional semver breaks,
    // we might want to assert this is !Send
}
// We can use a very similar trait to the one in the PR
trait AmbiguousIfImpl<A> { fn some_item(&self) {} }
impl<T: ?Sized> AmbiguousIfImpl<()> for T {}
struct Invalid;
impl<T: ?Sized + Send> AmbiguousIfImpl<Invalid> for T {}
// And use it to test the future is not Send
let fut = some_lib_feature(); // We can construct a value of the type, but can't name it
let () = <_ as AmbiguousIfImpl<_>>::some_item(&fut);
//        ^ type is deduced to the type of the future, although we can not name it. *
```

\* This comment might become a bit outdated with the stabilization of `type_alias_impl_trait` in rust-lang/rust#63063.
